frontend/src/main/java/com/google/tripmeout/frontend/storage/InMemoryPlaceVisitStorage.java

```diff
@@ -0,0 +1,103 @@
+package com.google.tripmeout.frontend.storage;
+
+import com.google.common.collect.HashBasedTable;
+import com.google.common.collect.Table;
+import com.google.common.collect.Tables;
+import com.google.tripmeout.frontend.PlaceVisitModel;
+import com.google.tripmeout.frontend.error.PlaceVisitAlreadyExistsException;
+import com.google.tripmeout.frontend.error.PlaceVisitNotFoundException;
+import com.google.tripmeout.frontend.error.TripNotFoundException;
+import com.google.tripmeout.frontend.storage.PlaceVisitStorage;
+import java.lang.RuntimeException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+
+public class InMemoryPlaceVisitStorage implements PlaceVisitStorage {
+  // <tripId, placeId, PlaceVisitModel>
+  Map<String, Map<String, PlaceVisitModel>> storage = new ConcurrentHashMap<>();
+
+  @Override
+  public void addPlaceVisit(PlaceVisitModel placeVisit)
+      throws PlaceVisitAlreadyExistsException, Throwable {
+    try {
+      storage.compute(placeVisit.tripId(), (tripKey, placesMap) -> {
+        if (placesMap == null) {
+          Map<String, PlaceVisitModel> newPlaceMap = new ConcurrentHashMap<>();
+          newPlaceMap.put(placeVisit.placeId(), placeVisit);
+          return newPlaceMap;
+        }
+        if (placesMap.get(placeVisit.placeId()) != null) {
+          throw new RuntimeException(new PlaceVisitAlreadyExistsException("PlaceVisit "
+              + placeVisit.name() + " already exists for trip " + placeVisit.tripId()));
+        }
+        placesMap.put(placeVisit.placeId(), placeVisit);
+        return placesMap;
+      });
+    } catch (RuntimeException e) {
+      throw e.getCause();
+    }
+  }
+
+  @Override
+  public void removePlaceVisit(String tripId, String placeId) throws PlaceVisitNotFoundException {
+    Map<String, PlaceVisitModel> placesMap = storage.get(tripId);
+    if (placesMap == null) {
+      throw new PlaceVisitNotFoundException(
+          "PlaceVisit with id" + placeId + " not found for trip " + tripId);
+    }
+
+    if (placesMap.remove(placeId) == null) {
+      throw new PlaceVisitNotFoundException(
+          "PlaceVisit with id" + placeId + " not found for trip " + tripId);
+    }
+  }
+
+  @Override
+  public Optional<PlaceVisitModel> getPlaceVisit(String tripId, String placeId) {
+    return Optional.ofNullable(storage.get(tripId)).map(placesMap -> placesMap.get(placeId));
+  }
+
+  @Override
+  public boolean changePlaceVisitStatus(String tripId, String placeId, String newStatus) {
+    Map<String, PlaceVisitModel> placesMap = storage.get(tripId);
+    if (placesMap == null) {
+      return false;
+    }
+
+    PlaceVisitModel originalPlace = placesMap.computeIfPresent(
```

Comment:
As above: don't you want to insert a new value if this isn't already present?

Reply:
I don't think I clearly thought this method through... Adding a new PlaceVisitModel object would make sense given the intent, but now that I've changed get to return an optional place visit object, I'm thinking now it might be better to just use get for checking whether the place is already in storage. do you think this might be better?

Reply:
Either approach is fine; we just want a consistent, reliable interface.

If you _do_ think you'll want to insert a new object at the same time as setting its status, then you'll need to actually update this method to do so.

If it's acceptable to require that visits be explicitly inserted before their statuses can be set, then this method should throw an exception in that case (rather than return a boolean).

Think about how callers will consume the interface. When they want to update a visit status, do you expect them to have to call `get` first to figure out if the object exists (and then insert the visit and finally update its status) or will they _know_ that the trip visit already exists? In the latter case, there's no reason to allow mutation in-line because callers are never interacting with non-existent visits. If callers need to query and conditionally insert updates in the common case, you'll probably want to support that here.